--- a/src/Makefile
+++ b/src/Makefile
@@ -10,6 +10,7 @@ CC= gcc -std=gnu99
 CFLAGS= -O2 -Wall -Wextra -DLUA_COMPAT_5_3 $(SYSCFLAGS) $(MYCFLAGS)
 LDFLAGS= $(SYSLDFLAGS) $(MYLDFLAGS)
 LIBS= -lm $(SYSLIBS) $(MYLIBS)
+LD= ld.lld

 AR= ar rcu
 RANLIB= ranlib
@@ -61,10 +62,22 @@ $(LUA_A): $(BASE_O)
 	$(RANLIB) $@

 $(LUA_T): $(LUA_O) $(LUA_A)
+ifeq ($(LD),ld.lld)
+	$(LD) -static --sysroot=$(SYSROOT) -L$(SYSROOT)/usr/lib -e _start \
+		$(SYSROOT)/usr/lib/crt0.o $(LUA_O) $(LUA_A) -o $@ \
+		-lm -lc $(SYSROOT)/usr/lib/crti.o $(SYSROOT)/usr/lib/crtn.o
+else
 	$(CC) -o $@ $(LDFLAGS) $(LUA_O) $(LUA_A) $(LIBS)
+endif

 $(LUAC_T): $(LUAC_O) $(LUA_A)
+ifeq ($(LD),ld.lld)
+	$(LD) -static --sysroot=$(SYSROOT) -L$(SYSROOT)/usr/lib -e _start \
+		$(SYSROOT)/usr/lib/crt0.o $(LUAC_O) $(LUA_A) -o $@ \
+		-lm -lc $(SYSROOT)/usr/lib/crti.o $(SYSROOT)/usr/lib/crtn.o
+else
 	$(CC) -o $@ $(LDFLAGS) $(LUAC_O) $(LUA_A) $(LIBS)
+endif

 test:
 	./$(LUA_T) -v