--- a/src/luaconf.h
+++ b/src/luaconf.h
@@ -90,6 +90,10 @@
 #define LUA_USE_LINUX
 #endif

+#if defined(__nightingale__)
+#define LUA_USE_POSIX
+#endif
+
 #if defined(LUA_USE_MACOSX)
 #define LUA_USE_POSIX
 #define LUA_DL_DYLD		/* MacOS does not need -ldl */
--- a/src/lua.c
+++ b/src/lua.c
@@ -40,6 +40,11 @@

 #if defined(LUA_USE_POSIX)   /* { */

+#if defined(__nightingale__)
+/* Include signal.h for Nightingale's posix_sigaction */
+#include <signal.h>
+#endif
+
 /*
 ** Use 'sigaction' when available.
 */
@@ -51,7 +56,11 @@ static void setsignal (int sig, void (*handler)(int)) {
   struct sigaction sa;
   sa.sa_handler = handler;
   sa.sa_flags = 0;
   sigemptyset(&sa.sa_mask);  /* do not mask any signal */
+#if defined(__nightingale__)
+  posix_sigaction(sig, &sa, NULL);
+#else
   sigaction(sig, &sa, NULL);
+#endif
 }

 #else           /* }{ */