diff --git a/config.sub b/config.sub
index 75bb6a3135b..5aa0aa9be4f 100755
--- a/config.sub
+++ b/config.sub
@@ -1363,7 +1363,8 @@ case $os in
 	     | powermax* | dnix* | nx6 | nx7 | sei* | dragonfly* \
 	     | skyos* | haiku* | rdos* | toppers* | drops* | es* \
 	     | onefs* | tirtos* | phoenix* | fuchsia* | redox* | bme* \
-	     | midnightbsd* | amdhsa* | unleashed* | emscripten*)
+	     | midnightbsd* | amdhsa* | unleashed* | emscripten* \
+         | nightingale*)
 	# Remember, each alternative MUST END IN *, to match a version number.
 		;;
 	qnx*)
diff --git a/fixincludes/mkfixinc.sh b/fixincludes/mkfixinc.sh
index 0f9648608e9..86de11cd989 100755
--- a/fixincludes/mkfixinc.sh
+++ b/fixincludes/mkfixinc.sh
@@ -11,6 +11,7 @@ target=fixinc.sh
 
 # Check for special fix rules for particular targets
 case $machine in
+    *-nightingale* | \
     i?86-*-cygwin* | \
     i?86-*-mingw32* | \
     x86_64-*-mingw32* | \
diff --git a/gcc/config.gcc b/gcc/config.gcc
index ddd3b8f4d9d..ed16bec8a65 100644
--- a/gcc/config.gcc
+++ b/gcc/config.gcc
@@ -1977,6 +1977,12 @@ x86_64-*-fuchsia*)
 	tmake_file="${tmake_file} i386/t-x86_64-elf"
 	tm_file="${tm_file} i386/unix.h i386/att.h elfos.h newlib-stdint.h i386/i386elf.h i386/x86-64.h fuchsia.h"
 	;;
+i[34567]86-*-nightingale*)
+    tm_file="${tm_file} i386/unix.h i386/att.h dbxelf.h elfos.h glibc-stdint.h i386/i386elf.h nightingale.h"
+    ;;
+x86_64-*-nightingale*)
+    tm_file="${tm_file} i386/unix.h i386/att.h dbxelf.h elfos.h glibc-stdint.h i386/i386elf.h i386/x86-64.h nightingale.h"
+    ;;
 ia64*-*-elf*)
 	tm_file="${tm_file} dbxelf.h elfos.h newlib-stdint.h ia64/sysv4.h ia64/elf.h"
 	tmake_file="ia64/t-ia64"
diff --git a/gcc/config/nightingale.h b/gcc/config/nightingale.h
new file mode 100644
index 00000000000..af16bb8a1af
--- /dev/null
+++ b/gcc/config/nightingale.h
@@ -0,0 +1,37 @@
+/* Useful if you wish to make target-specific GCC changes. */
+#undef TARGET_NIGHTINGALE
+#define TARGET_NIGHTINGALE 1
+ 
+/* Default arguments you want when running your
+   i686-nightingale-gcc/x86_64-nightingale-gcc toolchain */
+#undef LIB_SPEC
+#define LIB_SPEC "-lc" /* link against C standard library */
+ 
+#undef STARTFILE_SPEC
+#define STARTFILE_SPEC "crti.o%s crtbegin.o%s crt0.o%s"
+
+#undef ENDFILE_SPEC
+#define ENDFILE_SPEC "crtend.o%s crtn.o%s"
+
+/* Tell ld to force 4KB pages, and some other things */
+#undef LINK_SPEC
+#define LINK_SPEC \
+  "-z max-page-size=4096 \
+   %{assert*} %{R*} %{rpath*} \
+   %{shared} \
+   %{!shared: \
+     %{!static: \
+       %{rdynamic:-export-dynamic} \
+       -dynamic-linker /bin/ld-ng.so} \
+     %{static}}"
+
+/* Additional predefined macros. */
+#undef TARGET_OS_CPP_BUILTINS
+#define TARGET_OS_CPP_BUILTINS()      \
+  do {                                \
+    builtin_define ("__nightingale__");      \
+    builtin_define ("__unix__");      \
+    builtin_assert ("system=nightingale");   \
+    builtin_assert ("system=unix");   \
+    builtin_assert ("system=posix");   \
+  } while(0);
diff --git a/libgcc/config.host b/libgcc/config.host
index 91abc84da03..7c57a838bf2 100644
--- a/libgcc/config.host
+++ b/libgcc/config.host
@@ -665,6 +665,14 @@ i[34567]86-*-netbsdelf*)
 x86_64-*-netbsd*)
 	tmake_file="${tmake_file} i386/t-crtstuff"
 	;;
+i[34567]86-*-nightingale*)
+	extra_parts="$extra_parts crti.o crtbegin.o crtend.o crtn.o"
+	tmake_file="$tmake_file i386/t-crtstuff t-crtstuff-pic t-libgcc-pic"
+	;;
+x86_64-*-nightingale*)
+	extra_parts="$extra_parts crti.o crtbegin.o crtend.o crtn.o"
+	tmake_file="$tmake_file i386/t-crtstuff t-crtstuff-pic t-libgcc-pic"
+	;;
 i[34567]86-*-openbsd*)
 	;;
 x86_64-*-openbsd*)
